Allow ajax requests in Router's resolve function

dltmd6262 opened the pull request:
Related issue: #12

Commit:
Wrap Router's resolve function in Promise and resolve them in parallel with Promise.all
Router's open will now return a Promise with its run function that returns a destroyer
History will take this destroyer and set it after yielding the open function

omegascorp commented:
@dltmd6262 please check the unit tests, few rows in owl.history is not covered.

omegascorp commented:
@dltmd6262 it would be great if we will pass result of resolves to the controller action (or callback) as second parameter.

omegascorp commented on src/owl.Router.js:
```diff
@@ -26,39 +26,39 @@
+            return owl.Promise.all(resolves.map(function (resolve) {
+                const callback = owl.history.getResolve(resolve);
+                if (callback) {
+                    return owl.Promise.method(callback)();
```
owl.Promise.method is not included in Promise/A+ specification let's just call callback, and in case if result is not instanceof owl.Promise return owl.Promise.resolve(result);

omegascorp commented on src/owl.Router.js:
```diff
@@ -26,39 +26,39 @@
+                    return owl.Promise.method(callback)();
+                } else {
+                    console.info('Resolve' + resolve + 'is not found');
+                    return owl.Promise.resolve(true);
```
Let's resolve with null when resolve is not found.

omegascorp commented on src/owl.Router.js:
```diff
@@ -26,39 +26,39 @@
+                    console.info('Resolve' + resolve + 'is not found');
+                    return owl.Promise.resolve(true);
+                }
+            })).then(function (resultArray) {
```
This then is not needed let's call then and catch on this.resolve() method class.

omegascorp commented on src/owl.Router.js:
```diff
@@ -26,39 +26,39 @@
-                return this.run(path, route);
-            }
-            return null;
+            return this.resolve(route).then(function (ret) {
```
Let's name it resolveResult and pass it also this.run and then to action or callback as second parameter.

dltmd6262 commented:
Got it :] Please give me some time to look over your comments. As for the unit tests, I ran 'npm run test' and all the unit tests there seemed to pass. Are there any additional tests I should run before a PR? Thanks again!

omegascorp commented:
There should be also a coverage report generated in coverage/ folder. Find there index.html file it will tell you which rows are not covered by unit test.

Commit:
- Change Promise.method to Promise.resolve
- When no resolve callback is found in history, return null
- Check result of Router.resolve in Router.open and catch error there
- Name array of resolve results resolveResult and pass it to action and callback as second parameter
- increase test coverage

omegascorp commented on src/owl.Router.js:
```diff
@@ -34,40 +34,40 @@
-              return ret ? that.run(path, route) : null;
-            })
+            return this.resolve(route).then(function (resolveResult) {
+                const calledAllResolves = resolveResult.every(function (a) {return a;});
```
This is not needed if promise resolved with false or 0 we need to just pass it to the controller.action, action will decide what to do.

omegascorp commented on src/owl.Router.js:
```diff
@@ -34,40 +34,40 @@
+                const calledAllResolves = resolveResult.every(function (a) {return a;});
+                return calledAllResolves ? that.run(path, route, resolveResult) : null;
+            }).catch(function (e) {
+                console.error('Error in Router.open', e.message, e.stack);
```
Let's call defaultRoute here and pass error as second parameter to make possible to process errors.

omegascorp commented:
Another one small thing, there is typescript definitions in typescript/ folder. Property types which are changed in library need to be changed there too.

dltmd6262 commented:
Got it! Thanks for bearing with me. I'll fix it and make another commit :]

Commit:
Remove conditional in resolve that required all resolved values to be truthy
In try-catch call the defaultRoute with the error object
Modify typescript files to match the new function types

omegascorp commented:
@dltmd6262 cool, thank you for the commits :)

dltmd6262 commented:
Thank you for guiding me through the whole thing! Maybe I can contribute later on another issue :]

omegascorp commented:
Sure, you are welcome :)
